README.md
# Heptagon/BZR

Heptagon is a synchronous dataflow language whose syntax and semantics is
inspired from [Lustre](http://www-verimag.imag.fr/Synchrone,30.html),
with a syntax allowing the expression of control structures (e.g., switch or
mode automata).

Heptagon is also a research compiler, whose aim is to facilitate
experimentation. The current version of the compiler includes the following
features:
- **Inclusion of *discrete controller synthesis* within the
  compilation**: the language is equipped with a behavioral contract
  mechanisms, where assumptions can be described, as well as an "enforce"
  property part. The semantics of this latter is that the property should be
  enforced by controlling the behaviour of the node equipped with the
  contract. This property will be enforced by an automatically built controller,
  which will act on free controllable variables given by the programmer. This
  extension has been named [BZR](http://bzr.inria.fr) in previous
  works.
- **Expression and compilation of array values with modular memory
  optimization.** The language allows the expression and operations on
  arrays (access, modification, iterators). With the use of *location annotations*, the
  programmer can avoid unnecessary array copies.

  Heptagon is developed in
  the [Parkas (ENS)](http://www.di.ens.fr/ParkasTeam.html)
  and [Ctrl-A (LIG/INRIA)](http://team.inria.fr/ctrl-a) research teams.


## How to get it or try it

### Installation with OPAM

The easiest and recommended way to install Heptagon/BZR is to use
the OCaml Package Manager (OPAM).

The installation sequence using OPAM is :
1. Install OPAM (at least v2.0) : the procedure depends on your system and is
        described [on the OPAM webpage](http://opam.ocaml.org/doc/Install.html).
2. (optional, for the graphical simulator) Install
        the *gtk2.0* libraries (on debian systems, the package is
        named `libgtk2.0-dev`)
3. Initialize OPAM:
        ```
          opam init
        ```
        This command should install a version of OCaml at least ≥
        4.08. Follow the instructions given throughout this initialization.
4. Install Heptagon:
        ```opam install heptagon```

To use the controller synthesis tool ReaX with Heptagon/BZR :
1. Install the *mpfr* and *gmp* libraries (on debian
        systems, packages named `libmpfr-dev`
        and `libgmp-devel`)
2. Add the repository for ReaX and its libraries (named here
      nberth-repo) :
      ```
        opam repo add nberth-repo "http://nberth.space/opam-repo"
        opam update
      ```
3. Install ReaX and its libraries, and the BZReaX script:
      ```opam install bzreax```

The source code is also available from [the Heptagon/BZR repository](https://gitlab.inria.fr/synchrone/heptagon) for manual
      compilation and installation. Further indications about ReaX can
      be found on [the ReaX/ReaTk page](http://reatk.gforge.inria.fr).

### Manual installation
#### Download

Heptagon can be freely downloaded [here](https://gitlab.inria.fr/synchrone/heptagon).

#### Technical requirements

The use of the Heptagon compiler by itself does not require any
additional tools. However, the usual use involves a compiler for the
generated code (target languages are currently C or Java).

To manually compile and install the Heptagon compiler, the following tools and libraries are needed:
- ocamlfind
- The *menhir* tool
- The *ocamlgraph* library
- *camlp4*
- The *lablgtk* library (optional, for the graphical simulator)
- The *reatk* library (optional, for the backend towards the ReaX controller synthesis too)

The tools below are optional or are related to some subparts of Heptagon:
- The graphical display tool sim2chro can be obtained from
      [Verimag](https://www-verimag.imag.fr/The-Lustre-Toolbox.html). It can be used together with Heptagon's graphical simulator.

#### Compilation and installation

Once the previously described libraries and tools are installed, the Heptagon compiler and libraries can be installed with:
```
    ./configure
    make
    make install
```


CHANGES
Heptagon 1.05.00 (07/06/2018)
-----------------------------

 - Option -simple-scheduler active for scheduling post-ctrln code generation
 - Handle implication operator in Sigali backend
 - Compatibility with reatk >= 0.14

Heptagon 1.04.00 (29/09/2017)
-----------------------------

 - Lustre front-end (experimental) : syntax adaptations to parse Lustre programs
 - Bug fix in normalization of merge
 - Compatibility with reatk >= 0.11

Heptagon 1.03.04 (24/05/2017)
-----------------------------

 - Bug correction in init analysis (bug #14076)
 - Bug correction in causality analysis: contradictory dependencies
   in states of automata were accepted, non schedulable equations generated
 - optimization on several passes
 - added -simple-scheduler option, triggering very simple, time-efficient
   scheduling

Heptagon 1.03.03 (14/03/2017)
-----------------------------

 - Correction in Ctrl-n backend: suppression of erroneous "c_" prefixes
 - Correction in Ctrl2ept: no systematic conversion from int to float
 - Correction in Ctrl-n backend: handling of implication and "xor" operators

Heptagon 1.03.02 (08/02/2017)
-----------------------------

 - Added operator '=>' (logical implication)
 - Systematic sink state in Ctrl-n backend, outputs as local variables

Heptagon 1.03.01 (02/06/2016)
-----------------------------

 - Correct typing of numericals when translating Controllable-Nbac nodes
 - Handling of several controllers by module and correct modular controller synthesis
 - Correction of Java target with controllers

Heptagon 1.03.00 (18/01/2016)
-----------------------------

 - -deadcode option, improvement of deadcode removal
 - graphical simulator handles type aliasing

Heptagon 1.02.00 (13/12/2015)
-----------------------------

 - syntax for attractivity and reachability in contracts
 - option to force abstraction of infinite-domain state variables (ctrl-n)

Heptagon 1.01.00 (17/09/2015)
-----------------------------

 - back-end towards controller synthesis tool ReaX

Heptagon 1.00.06 (21/02/2014)
-----------------------------

 - allowed abstractions (non-inlined subnodes calls) in contracts

Heptagon 1.00.05 (28/01/2014)
-----------------------------

 - bug fix: correct handling of nodes with contracts inside automata and switches

Heptagon 1.00.04 (14/01/2014)
-----------------------------

 - bug fix: comparison between two non-constant integer expressions in Sigali
 - bug fix: correct handling of "=" and "<>" operators in Sigali

Heptagon 1.00.03 (20/11/2013)
-----------------------------

 - bug fix: tomato application with contracts

Heptagon 1.00.02 (29/10/2013)
-----------------------------

 - scrollbars in graphical simulator hepts

Heptagon 1.00.01 (05/08/2013)
-----------------------------

 - bug fix : abstraction of integer in Sigali code generation


web/index.php
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN"
                      "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
<html>
  <head>
    <link type="text/css" href="style.css" rel="stylesheet">
    <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
    <title>Heptagon/BZR</title>
  </head>
  
<body>
  
<?php include('toc.php'); ?>

<?php include('header.php'); ?>

<div id="content">
<p>
Heptagon is a synchronous dataflow language whose syntax and semantics is
inspired from <a href="http://www-verimag.imag.fr/Synchrone,30.html">Lustre</a>,
with a syntax allowing the expression of control structures (e.g., switch or
mode automata).
</p>
<p>
Heptagon is also a research compiler, whose aim is to facilitate
experimentation. The current version of the compiler includes the following
features:
<ul>
  <li><strong>Inclusion of <em>discrete controller synthesis</em> within the
  compilation</strong>: the language is equipped with a behavioral contract
  mechanisms, where assumptions can be described, as well as an "enforce"
  property part. The semantics of this latter is that the property should be
  enforced by controlling the behaviour of the node equipped with the
  contract. This property will be enforced by an automatically built controller,
  which will act on free controllable variables given by the programmer. This
  extension has been named <a href="http://bzr.inria.fr">BZR</a> in previous
  works.</li>
  <li><strong>Expression and compilation of array values with modular memory
  optimization.</strong> The language allows the expression and operations on
  arrays (access, modification, iterators). With the use of <em>location annotations</em>, the
  programmer can avoid unnecessary array copies.</li>
</ul>
</p>

<p>
  Heptagon is developed in
  the <a href="http://www.di.ens.fr/ParkasTeam.html">Parkas (ENS)</a>
  and <a href="http://team.inria.fr/ctrl-a">Ctrl-A (LIG/INRIA)</a> research teams.
</p>

<h2>How to get it or try it</h2>

<h3>Installation with OPAM</h3>

<p>The easiest and recommended way to install Heptagon/BZR is to use
the OCaml Package Manager (OPAM).</p>

<p>The installation sequence using OPAM is :
    <ol>
      <li>Install OPAM : the procedure depends on your system and is
        described <a href="http://opam.ocaml.org/doc/Install.html">on
          the OPAM webpage</a>.</li>
      <li>(optional, for the graphical simulator) Install
        the <i>gtk2.0</i> libraries (on debian systems, the package is
        named <tt>libgtk2.0-dev</tt>)</li>
      <li>Initialize OPAM:
        <pre>
          opam init
          opam switch 4.02.3
          eval `opam config env`
        </pre>
      </li>
      <li>Install Heptagon:
        <pre>opam install heptagon</pre>
      </li>
    </ol>
</p>
<p>To use the controller synthesis tool ReaX with Heptagon/BZR :
  <ol>
    <li>Install the <i>mpfr</i> and <i>gmp</i> libraries (on debian
        systems, packages named <tt>libmpfr-devel</tt>
        and <tt>libgmp-devel</tt>)</li>
    <li>Add the repository for ReaX and its libraries (named here
      nberth-repo) :
      <pre>
        opam repo add nberth-repo "http://nberth.space/opam-repo"
        opam update
      </pre>
    </li>
    <li>Install ReaX and its libraries, and the BZReaX script:
      <pre>opam install fixpoint bddapron reatk bzreax</pre>
    </li>
  </ol>
</p>
<p>The source code is also available from <a
        href="http://heptagon.gforge.inria.fr">the Heptagon/BZR web page</a> for manual
      compilation and installation. Further indications about ReaX can
      be found on <a href="http://reatk.gforge.inria.fr">the ReaX/ReaTk
        page</a>.<br>
    </p>

<h3>Manual installation</h3>
<h4>Download</h4>

<p>
Heptagon can be freely downloaded <a href="http://gforge.inria.fr/projects/heptagon">here</a>.
</p>

<h4>Technical requirements</h4>
<p>  
The use of the Heptagon compiler by itself does not require any
additional tools. However, the usual use involves a compiler for the
generated code (target languages are currently C or Java).
</p>
<p>
  To manually compile and install the Heptagon compiler, the following tools and libraries are needed:
  <ul>
    <li>ocamlfind</li>
    <li>The <em>menhir</em> tool</li>
    <li>The <em>ocamlgraph</em> library</li>
    <li><em>camlp4</em></li>
    <li>The <em>lablgtk</em> library (optional, for the graphical simulator)</li>
    <li>The <em>reatk</em> library (optional, for the backend towards the ReaX controller synthesis too)</li>
  </ul>
</p>
<p>
The tools below are optional or are related to some subparts of Heptagon:
<ul> 
  <li>The graphical display tool sim2chro can be obtained from
    <a href="http://www-verimag.imag.fr/~raymond/edu/distrib/">
      Verimag</a>. It can be used together with Heptagon's graphical simulator.</li>
  <li> <a href="https://gforge.inria.fr/projects/bzr">Sigali</a>, the
    controller synthesis tool, developed by the Espresso and Vertecs team at INRIA
    Rennes. </li>
</ul>
</p>

<h4>Compilation and installation</h4>

<p>Once the previously described libraries and tools are installed, the Heptagon compiler and libraries can be installed with:
  <pre>
    ./configure
    make
    make install
  </pre>
</p>

<h2>Contact</h2>

<p>
Please
contact <a href="mailto:heptagon-developers@lists.gforge.inria.fr">us</a>
for further information.
</p>

<h2>Main participants</h2>

<table>
  <tr>
    <td>Gwena&euml;l Delaval</td>
    <td>Assistant Prof. at <a href="http://www.univ-grenoble-alpes.fr/">UGA</a></td>
    <td>+33 4 76 61 54 31</td>
    <td><a href="mailto:gwenael.delaval@inria.fr">mail</a></td>
    <td><a href="http://pop-art.inrialpes.fr/people/delaval/">web</a></td>
  </tr>
  <tr>
    <td>Herv&eacute; Marchand</td>
    <td>Researcher at <a href="http://www.inria.fr/">INRIA</a></td>
    <td>+33 2 99 84 75 09</td>
    <td><a href="mailto:herve.marchand@inria.fr">mail</a></td>
    <td><a href="http://www.irisa.fr/prive/hmarchan/">web</a></td>
  </tr>
  <tr>
    <td>Marc Pouzet</td>
    <td>Professor at <a href="http://www.ens.fr/">ENS</a></td>
    <td></td>
    <td><a href="mailto:marc dot pouzet at ens dot fr">mail</a></td>
    <td><a href="http://www.di.ens.fr/~pouzet/">web</a></td>
  </tr>
  <tr>
    <td>Eric Rutten</td>
    <td>Researcher at <a href="http://www.inria.fr/">INRIA</a></td>
    <td>+33 4 76 61 55 50</td>
    <td><a href="mailto:eric.rutten@inria.fr">mail</a></td>
    <td><a href="http://sardes.inrialpes.fr/~rutten">web</a></td>
  </tr>
</table>
</div>
</body>
</html>
